osu.Game/Rulesets/UI/HitObjectContainer.cs

```diff
@@ -1,78 +1,174 @@
-            startTimeMap[hitObject].bindable.BindValueChanged(_ => onStartTimeChanged(hitObject));
+            Debug.Assert(!drawableMap.ContainsKey(entry));
+
+            var drawable = drawableRuleset.GetDrawableRepresentation(entry.HitObject);
+            drawable.OnNewResult += onNewResult;
+            drawable.OnRevertResult += onRevertResult;
+
+            bindStartTime(drawable);
+            AddInternal(drawableMap[entry] = drawable, false);
+        }
+
+        private void removeDrawable(HitObjectLifetimeEntry entry)
+        {
+            Debug.Assert(drawableMap.ContainsKey(entry));
+
+            var drawable = drawableMap[entry];
+            drawable.OnNewResult -= onNewResult;
+            drawable.OnRevertResult -= onRevertResult;
+            drawable.OnKilled();
+
+            drawableMap.Remove(entry);
 
+            unbindStartTime(drawable);
+            RemoveInternal(drawable);
+        }
+
+        #endregion
+
+        #region Non-pooling support
+
+        public virtual void Add(DrawableHitObject hitObject)
+        {
+            bindStartTime(hitObject);
             AddInternal(hitObject);
+
+            hitObject.OnNewResult += onNewResult;
+            hitObject.OnRevertResult += onRevertResult;
         }
 
         public virtual bool Remove(DrawableHitObject hitObject)
         {
             if (!RemoveInternal(hitObject))
                 return false;
 
-            // Removed last for the comparer to remain ordered during RemoveInternal
-            startTimeMap[hitObject].bindable.UnbindAll();
-            startTimeMap.Remove(hitObject);
+            hitObject.OnNewResult -= onNewResult;
+            hitObject.OnRevertResult -= onRevertResult;
+
+            unbindStartTime(hitObject);
 
             return true;
         }
 
-        protected override void Dispose(bool isDisposing)
+        public int IndexOf(DrawableHitObject hitObject) => IndexOfInternal(hitObject);
+
+        protected override void OnChildLifetimeBoundaryCrossed(LifetimeBoundaryCrossedEvent e)
         {
-            base.Dispose(isDisposing);
-            unbindStartTimeMap();
+            if (!(e.Child is DrawableHitObject hitObject))
+                return;
+
+            if ((e.Kind == LifetimeBoundaryKind.End && e.Direction == LifetimeBoundaryCrossingDirection.Forward)
+                || (e.Kind == LifetimeBoundaryKind.Start && e.Direction == LifetimeBoundaryCrossingDirection.Backward))
+            {
+                hitObject.OnKilled();
+            }
         }
 
+        #endregion
+
         public virtual void Clear(bool disposeChildren = true)
         {
+            lifetimeManager.ClearEntries();
+
             ClearInternal(disposeChildren);
-            unbindStartTimeMap();
+            unbindAllStartTimes();
         }
 
-        private void unbindStartTimeMap()
+        protected override bool CheckChildrenLife() => base.CheckChildrenLife() | lifetimeManager.Update(Time.Current, Time.Current);
```

Comment:
I don't really like code like this (would prefer to have a three-liner method instead, which I assume was the intention here) but not sure if it's just me.

Reply:
Just so I understand, this is what you're suggesting?:

![image](https://user-images.githubusercontent.com/1329837/98893648-3d08d880-24e6-11eb-832a-0252f74b9e36.png)

Reply:
Yeah, that's what I had in mind.